### PR TITLE
[Investigating - Don't merge] Bullet visuals, better pen, shotgun debuff

### PR DESCRIPTION
**Pull request by AdamTadeusz**
## Description

A toggle to draw the path bullets take

cl_neo_bullet_trace 1 to see where bullets are going
cl_neo_bullet_trace_max_pen controls the colour of the bullet trace. The closer a bullet's pen is to this value the more red it will be. The further away it is, the more yellow it will be

Fixes a problem with bullet penetration where if multiple thin materials were stacked closely together, some could be missed completely by the bullet as it would continue along its trajectory further ahead. Also now when a player entity is hit by a bullet, it is added to an ignore filter, so that bullet can never hit that player entity again.

Also changes how shotguns pellets work, so all pellets are now tracelines that effectively have no width and can only hit whatever is directly in front of them, as opposed to the previous behaviour where half of the pellets behaved like a cube 6hu*6hu*6hu large, which are much more likely to hit something.

## Toolchain

- Windows MSVC VS2022
- Linux GCC Distro Native [Specify distro + GCC version]
- Linux GCC 10 Sniper 3.0

### CODE REVIEW

File: mp/src/game/shared/baseentity_shared.cpp [hunk 1]
```diff
@@ -2091,12 +2115,18 @@ void CBaseEntity::FireBullets( const FireBulletsInfo_t &info )
...
 
 		// See if we should pass through glass
+#ifndef NEO
 #ifdef GAME_DLL
 		if ( bHitGlass )
 		{
 			HandleShotImpactingGlass( info, tr, vecDir, &traceFilter );
 		}
 #endif
+#endif
```
Comment (brysondev) [inline mp/src/game/shared/baseentity_shared.cpp:2125]: Nit:
```suggestion
#endif // NEO
```

File: mp/src/game/shared/baseentity_shared.cpp [hunk 2]
```diff
@@ -2091,12 +2115,18 @@ void CBaseEntity::FireBullets( const FireBulletsInfo_t &info )
 		//NOTENOTE: We could expand this to a more general solution for various material penetration types (wood, thin metal, etc)
 
 		// See if we should pass through glass
+#ifndef NEO
 #ifdef GAME_DLL
 		if ( bHitGlass )
 		{
 			HandleShotImpactingGlass( info, tr, vecDir, &traceFilter );
 		}
 #endif
```
Comment (brysondev) [inline mp/src/game/shared/baseentity_shared.cpp:2124]: Nit: 
```suggestion
#endif // GAME_DLL
```

**Review by brysondev — APPROVED** [review]
idk if we care enough about `#endif`'s `// NEO` that are only a few lines but I nit marked a few that were more spaced out.

File: mp/src/game/shared/baseentity_shared.cpp [hunk 3]
```diff
@@ -1695,10 +1701,12 @@ void CBaseEntity::FireBullets( const FireBulletsInfo_t &info )
 	Vector vecDir;
 	Vector vecEnd;
 	
+#ifndef NEO // Each bullet get's its own filter instead, so each individual pellet can hit the same target
 	// Skip multiple entities when tracing
 	CBulletsTraceFilter traceFilter( COLLISION_GROUP_NONE );
 	traceFilter.SetPassEntity( this ); // Standard pass entity for THIS so that it can be easily removed from the list after passing through a portal
 	traceFilter.AddEntityToIgnore( info.m_pAdditionalIgnoreEnt );
+#endif
```
Comment (brysondev) [inline mp/src/game/shared/baseentity_shared.cpp]: https://github.com/user-attachments/assets/c928d571-407d-4e32-bdd1-f475b5016564

**Review by brysondev — APPROVED** [review]

**Review by VantoNortim — APPROVED** [review]

File: mp/src/game/shared/baseentity_shared.cpp [hunk 4]
```diff
@@ -2034,10 +2056,8 @@ void CBaseEntity::FireBullets( const FireBulletsInfo_t &info )
 				}
 			}
 #endif
-			// Bullet penetration
-			if (!bHitGlass)
-				HandleShotPenetration(info, tr, vecDir, &traceFilter);
 		}
+#endif // NEO
```
[the remark this answers is not in the document]
Comment (AdamTadeusz) [reply mp/src/game/shared/baseentity_shared.cpp:2064]: This is not original engine code

File: mp/src/game/shared/baseentity_shared.cpp [hunk 5]
```diff
@@ -2170,18 +2196,36 @@ void CBaseEntity::HandleShotPenetration(const FireBulletsInfo_t& info,
...
 
 	data.m_vNormal = tr.plane.normal;
 	data.m_vOrigin = tr.endpos;
 
 	trace_t	penetrationTrace;
 
-	// Re-trace as if the bullet had passed right through
-	UTIL_TraceLine(testPos, tr.endpos, MASK_SHOT, pTraceFilter, &penetrationTrace);
+	// Re-trace backwards to find the bullet ext
+	UTIL_TraceLine(testPos, tr.endpos, MASK_SHOT, nullptr, &penetrationTrace);
```
[the remark this answers is not in the document]
Comment (AdamTadeusz) [reply mp/src/game/shared/baseentity_shared.cpp:2235]: ditto

File: mp/src/game/shared/baseentity_shared.cpp [hunk 6]
```diff
@@ -2215,6 +2259,16 @@ void CBaseEntity::HandleShotPenetration(const FireBulletsInfo_t& info,
...
+	if (cl_neo_bullet_trace.GetBool())
+	{ // Exit (RED STAR)
+		Vector x0 = penetrationTrace.endpos + Vector(-0.5, 0, 0);
+		Vector x1 = penetrationTrace.endpos + Vector(0, -0.5, 0);
+		Vector x2 = penetrationTrace.endpos + Vector(0, 0, -0.5);
+		DebugDrawLine(x0, x0 + Vector(1, 0, 0), 0, 255, 0, 1, 30.f);
+		DebugDrawLine(x1, x1 + Vector(0, 1, 0), 0, 255, 0, 1, 30.f);
+		DebugDrawLine(x2, x2 + Vector(0, 0, 1), 0, 255, 0, 1, 30.f);
+	}
+
```
[the remark this answers is not in the document]
Comment (AdamTadeusz) [reply mp/src/game/shared/baseentity_shared.cpp]: same as above

**Review by Rainyan — APPROVED** [review]
This might be somewhat unrelated, but it appears the `FireBullets` is sometimes called multiple times for a single gunshot:

nt_dimensions_test0001

All 3 bullet holes in this screenshot were from 3 different mouse clicks, but for some reason the middle one has 4 traces on it (expected to see only 1). This multiple-lines-rendering behaviour seems to also happen quite often if you shoot in the far distance with the SRS.

**Comment by AdamTadeusz** [comment]
> This might be somewhat unrelated, but it appears the `FireBullets` is sometimes called multiple times for a single gunshot:
> 
> nt_dimensions_test0001
> 
> All 3 bullet holes in this screenshot were from 3 different mouse clicks, but for some reason the middle one has 4 traces on it. This multiple lines rendering seems to also happen quite often if you shoot in the far distance with the SRS.

I did make a change so that the firing of penetrating bullets is done on both client AND server. This is different to the original functionality of bullets penetrating glass for example (which is the only material that has penetration in the base sdk) where the initial bullet is fired by both client and server, but the bullet fired behind glass is only done on the server, which meant no bullet impacts were being drawn on the second wall that the penetrating bullet hit. I did some testing firing a bullet only client side and not server side and observer no actual damage being done to the player, but I will look into it again.

Also if this is only with the srs then it probably is something unrelated.

**Comment by Rainyan** [comment]
> Also if this is only with the srs then it probably is something unrelated.

I saw this with both the SRS and the Tachi, but I don't think it's specific to a particular gun. For whatever reason, shooting the SRS down the long end of the dimensions test map seemed to reproduce it the most often, but could be inconsequential.

**Comment by AdamTadeusz** [comment]
> > Also if this is only with the srs then it probably is something unrelated.
> 
> This particular screenshot above is from 3 bullets from with the Tachi in semi-auto mode, to produce the 3 different bullet visible holes in the wall. I saw this happening with both the SRS and the Tachi, but I don't think it's specific to a particular gun (I didn't test all the guns). For whatever reason, shooting the SRS down the long end of the dimensions test map seemed to reproduce it the most often, but could be inconsequential.

seems that the value of m_flNextPrimaryAttack is being reset, so itemPostFrame triggers PrimaryAttack() again. Removing the piece of code right above PrimaryAttack() in neobasecombatweapon::ItemPostFrame that does exactly that somehow only seems to make the problem worse. I don't think this PR is to blame, probably we're missing some m_flNextPrimaryAttack = max(m_flNextPrimaryAttack, calculated_m_flNextPrimaryAttack)

**Pull request merged**